# Introduction to Data Science 2025

# Week 1

## Exercise 1 | Matrix warm-up
<span style="background-color: #ccfff2"> *Note: You can find tutorials for NumPy and Pandas under 'Useful tutorials' in the course material.*</span>

One of the most useful properties of any scientific programming language (Python with NumPy, R, Julia, etc) is that they allow us to work with matrices efficiently. Let's learn more about these features!

### 1.1 Basics

1. Let's start by creating two arrays <span style="background-color: #ccfff2"> A</span> and <span style="background-color: #ccfff2"> B</span> which each have the integers <span style="background-color: #ccfff2"> 0, 1, 2, ..., 1e7-1</span>. Use the normal arrays or lists of the programming language you are using, e.g. *list* or *[ ]* or *numpy.array()* in Python.

In [1]:
import numpy as np
A = list(range(int(1e7)))
B = list(range(int(1e7)))

2. Create a function that uses a <span style="background-color: #ccfff2"> for loop</span> or equivalent to return a new array <span style="background-color: #ccfff2"> C</span>, which contains the <span style="background-color: #ccfff2"> element-wise sum of *A* and *B*</span>, e.g. C should contain the integers <span style="background-color: #ccfff2"> 0, 2, 4, etc</span>.

In [2]:
def add_with_for(A,B):
    C = []
    for i in range(len(A)):
        C.append(A[i] + B[i])
    return C

3. Next, let's create another function that uses NumPy (or equivalent) to do the same. To try it out, allocate two arrays (e.g. using <span style="background-color: #ccfff2"> np.array</span> in NumPy) and add the arrays together using your function. Don't use loops, instead, find out how to add the two arrays directly. What do you notice in comparison to the previous function?

In [3]:
def add_direct(A,B):
    A_numpy = np.array(A)
    B_numpy = np.array(B)
    C_numpy = A_numpy + B_numpy
    return C_numpy.tolist()

### 1.2 Array manipulation

<span style="background-color: #ccfff2"> *Note: for the following exercises, only use NumPy or equivalent functions. Don't use any loops.* </span>
1. Create the following array:

*[hint: <span style="background-color: #ccfff2"> np.reshape</span>]*

In [4]:
# array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
#        [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
#        [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
#        [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
#        [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
#        [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
#        [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
#        [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
#        [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
#        [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])

In [5]:
array = np.arange(100).reshape(10, 10)
print(array)

[[ 0  1  2  3  4  5  6  7  8  9]
 [10 11 12 13 14 15 16 17 18 19]
 [20 21 22 23 24 25 26 27 28 29]
 [30 31 32 33 34 35 36 37 38 39]
 [40 41 42 43 44 45 46 47 48 49]
 [50 51 52 53 54 55 56 57 58 59]
 [60 61 62 63 64 65 66 67 68 69]
 [70 71 72 73 74 75 76 77 78 79]
 [80 81 82 83 84 85 86 87 88 89]
 [90 91 92 93 94 95 96 97 98 99]]


2. Create the following array:

In [6]:
# array([[0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.]])

In [7]:
array = np.tile([0., 1.],(10,5))
print(array)

[[0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]
 [0. 1. 0. 1. 0. 1. 0. 1. 0. 1.]]


3. Create the following array (D):

In [8]:
# array([[0., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 0., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 0., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 0., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 0., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 0., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 0., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 0., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 1., 0.]])

In [9]:
D = np.ones((10,10))
np.fill_diagonal(D, 0)
print(D) 

[[0. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 0. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 0. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 0. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 0. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 0. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 0. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 0.]]


4. Create the following array (E):

In [10]:
# array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 0., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 0., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 0., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 0., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 0., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 0., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 0., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [0., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

In [11]:
E = np.fliplr(D)
print(E) 

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 0.]
 [1. 1. 1. 1. 1. 1. 1. 1. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 0. 1. 1.]
 [1. 1. 1. 1. 1. 1. 0. 1. 1. 1.]
 [1. 1. 1. 1. 1. 0. 1. 1. 1. 1.]
 [1. 1. 1. 1. 0. 1. 1. 1. 1. 1.]
 [1. 1. 1. 0. 1. 1. 1. 1. 1. 1.]
 [1. 1. 0. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 1. 1. 1. 1. 1. 1. 1. 1.]
 [0. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


5. Call the last two matrices <span style="background-color: #ccfff2">D</span> and <span style="background-color: #ccfff2">E</span>, respectively. Show that the determinant of their product (matrix multiplication) is the same as the product of their determinants. That is calculate both <span style="background-color: #ccfff2">det(DE)</span> and <span style="background-color: #ccfff2">det(D) * det(E)</span>, and show that they are the same. Is it a coincidence? (I think not) The product of the determinants (or the determinant of the product) should be -81.

In [12]:
DE = np.dot(D, E)
det_DE = np.linalg.det(DE)
det_D = np.linalg.det(D)
det_E = np.linalg.det(E)
print(det_D * det_E)
print(det_DE)

-81.0
-81.00000000000003


This is not a coincidence. By the property of determinants, we know this well held, and the only difference between -81.0 and -81.00000000000003 is due to floating-point computation errors by computer.


### 1.3 Slicing

Array slicing is a powerful way to extract data from an array. Let's practice array slicing with the following exercises!

1. Load the [California housing dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_california_housing.html). The data should be a matrix of shape <span style="background-color: #ccfff2">(20640, 8)</span>, that is 20640 rows and 8 columns. Use the <span style="background-color: #ccfff2">.shape</span> attribute of NumPy arrays to verify this. Here's a [description of the fields](https://scikit-learn.org/stable/datasets/real_world.html#california-housing-dataset).

In [13]:
from sklearn.datasets import fetch_california_housing
data = fetch_california_housing()
print(np.shape(data.data))
list(data.feature_names) 

(20640, 8)


['MedInc',
 'HouseAge',
 'AveRooms',
 'AveBedrms',
 'Population',
 'AveOccup',
 'Latitude',
 'Longitude']

2. Select rows where the average number of bedrooms <span style="background-color: #ccfff2">(AveBedrms)</span> is higher than 2. The first few row indices should be <span style="background-color: #ccfff2">710,  1023,  1024, ...</span> (zero-indexed). Count these houses - how many rows are selected? *[hint: <span style="background-color: #ccfff2">np.where</span>]*

In [14]:
idx = np.where(data.data[:,3] > 2)[0]
print(idx)
np.size(idx)# roles selected

[  710  1023  1024  1030  1102  1233  1234  1235  1236  1238  1239  1240
  1566  1867  1871  1872  1879  1881  1889  1901  1906  1910  1911  1912
  1913  1914  1925  1926  1978  1979  2392  2395  2396  2397  2398  2510
  2511  2643  2656  2763  2768  2776  2778  3014  3086  3095  3258  3292
  3293  3313  3314  3334  3349  3350  9193  9423  9431  9451  9634  9669
  9671  9672  9675  9676  9678  9679  9680  9681  9682  9683 10067 10076
 10077 10078 10079 10080 10081 10082 10083 10084 10405 11705 11706 11707
 11708 11709 11710 11711 11713 11714 11715 11716 11717 11719 11720 11721
 11722 11723 11724 11725 11729 11831 11832 11833 11834 11847 11848 11849
 11850 11862 11865 11866 11867 11869 11870 12136 12303 12305 12306 12307
 12324 12325 12344 12349 12351 12354 12359 12361 12362 12366 12368 12371
 12372 12374 12376 12390 12392 12394 12396 12404 12405 12430 12446 12447
 13898 13900 13911 13912 13919 13920 13923 13924 13934 13935 13936 13937
 13939 13940 13941 13942 13943 13944 13945 13946 13

235

3. Select the rows where the median house age (i.e. median in each block group) <span style="background-color: #ccfff2">(HouseAge)</span> is between 1 and 3 years (inclusive). There should be **124** of these.

In [15]:
idx = np.where((data.data[:,1] >= 1) & (data.data[:,1] <= 3))[0]
print(idx)
np.size(idx)# roles selected

[   59   854   871   959   976  1260  1565  1566  1648  1719  2206  2310
  2311  2312  2336  2339  2706  2774  2993  3130  3140  5506  9109  9132
  9147  9168  9187  9378  9639 10336 10376 10391 10401 10403 10406 10410
 10412 10416 10440 10456 10467 10513 10514 10519 10521 10524 10525 10529
 10548 10619 10622 10631 11490 11775 12000 12006 12077 12084 12119 12123
 12137 12142 12143 12151 12156 12167 12171 12201 12202 12203 12204 12285
 12286 12308 12363 12430 12841 12868 13041 13074 13100 13102 13139 13177
 13179 13367 13368 13373 13374 13375 13422 13659 13713 13715 13747 13848
 13855 14557 15220 15444 15464 15736 15785 16383 16566 16593 17068 17824
 17826 17827 18725 18942 18947 18972 18989 19001 19003 19033 19046 19536
 19561 19650 20070 20367]


124

4. Find the mean of the block group population <span style="background-color: #ccfff2">(Population)</span> for homes whose median value is more than 25000 USD (the target variable). It should be around **1425.68**.

In [16]:
population = np.where(data.data[:,4] > 2.5)[0]
mean_population = np.mean(data.data[population,4])
print(mean_population)


1425.4767441860465


## Exercise 2 | Working with text data

Next, let's look into some text data. We will be looking into Amazon reviews, and the necessary steps to transform a raw dataset into a format more suitable for prediction tasks.

1. Download the automotive 5-core dataset from [here](https://jmcauley.ucsd.edu/data/amazon_v2/categoryFilesSmall/Automotive_5.json.gz). Next, you can extract the data in <span style="background-color: #ccfff2">JSON</span> format. You can also download one of the bigger ones, if you are feeling ambitious. Open the JSON file. Access the <span style="background-color: #ccfff2">reviewText</span> field, which contains the unstructured review text written by the user.

For instance, the first review reads as follows: 

*'After I wrote the below review, the manufacturer contacted me and explained how to use this.  Instead of the (current) picture on Amazon where the phone is placed vertically, you actually use the stand with the phone placed horizontally. [...]'*

In [17]:

import json
import pandas as pd
df = pd.read_json('Automotive_5.json', lines=True)
print(df.columns)
review_texts = df['reviewText']
print(review_texts)

Index(['overall', 'verified', 'reviewTime', 'reviewerID', 'asin', 'style',
       'reviewerName', 'reviewText', 'summary', 'unixReviewTime', 'vote',
       'image'],
      dtype='object')
0          After I wrote the below review, the manufactur...
1          It sucks barely picks up anything definitely n...
2          Well to write a short one, it blew 2 fuses of ...
3          I have absolutely no memory of buying this but...
4                                       it ok it does it job
                                 ...                        
1711514                          Fast Shipping, Works Great!
1711515    Fit isn't great, relays are very hard to press...
1711516    I bought an auxiliary horn for my car and it n...
1711517    Way better than stock. Nice bright and white. ...
1711518    Great product fast shipping and awesome commun...
Name: reviewText, Length: 1711519, dtype: object


2. Next, let's follow some steps to normalize the text data.

When dealing with natural language, it is important to notice that while, for example, the words "Copper" and "copper" are represented by two different strings, they have the same meaning. When applying statistical methods on this data, it is useful to ensure that words with the same meaning are represented by the same string.

* <span style="background-color: #ccfff2">Downcasing</span>: Let's first downcase the contents of the <span style="background-color: #ccfff2">reviewText</span> field.

Now the first review should be:

*'after i wrote the below review, the manufacturer contacted me and explained how to use this.  instead of the (current) picture on amazon where the phone is placed vertically, you actually use the stand with the phone placed horizontally.'*

In [18]:
texts_lower = df['reviewText'].str.lower()
print(texts_lower)


0          after i wrote the below review, the manufactur...
1          it sucks barely picks up anything definitely n...
2          well to write a short one, it blew 2 fuses of ...
3          i have absolutely no memory of buying this but...
4                                       it ok it does it job
                                 ...                        
1711514                          fast shipping, works great!
1711515    fit isn't great, relays are very hard to press...
1711516    i bought an auxiliary horn for my car and it n...
1711517    way better than stock. nice bright and white. ...
1711518    great product fast shipping and awesome commun...
Name: reviewText, Length: 1711519, dtype: object


3. Let's continue with punctuation and stop word removal. Stop words are words like "and", "the", etc. They are usually very common words that have little to do with the actual content matter. There's plenty openly available lists of stop words for almost any (natural) language.

* <span style="background-color: #ccfff2">Punctuation and stop-word removal</span>: Let's now remove all punctuation, as well as the stop words. You can find a stop word list for English, e.g. [here](https://gist.github.com/xldrkp/4a3b1a33f10d37bedbe0068f2b4482e8#file-stopwords-en-txt).*(use the link to download a txt of english stopwords)* Save the stopwords in the file as "stopwords-en.txt".

First review at this point reads as: 

*'wrote review manufacturer contacted explained current picture amazon phone vertically stand phone horizontally'*

In [19]:
import string
import re
with open('stopwords-en.txt', 'r') as file:
      stopwords = set(file.read().strip().split('\n'))
def remove_punctuation(text):
    text = str(text)
    text = re.sub(r'[^\w\s]', ' ',text)
    words = [word for word in text.split() if word not in stopwords]
    return ' '.join(words)
texts_cleaned = texts_lower.apply(remove_punctuation)
print(texts_cleaned)



0          wrote review manufacturer contacted explained ...
1          sucks barely picks definitely cars pretty don ...
2          write short blew 2 fuses car apparently demand...
3          absolutely memory buying m going review goes a...
4                                                ok does job
                                 ...                        
1711514                            fast shipping works great
1711515    fit isn t great relays hard press connector me...
1711516    bought auxiliary horn car needed run relay guy...
1711517    way better stock nice bright white super easy ...
1711518    great product fast shipping awesome communicat...
Name: reviewText, Length: 1711519, dtype: object


4. Let's continue with stemming. For example, while the words "swims" and "swim" are different strings, they both refer to swimming. [Stemming](https://en.wikipedia.org/wiki/Stemming) refers to the process of mapping words from their inflected form to their base form, for instance: swims -> swim.

* <span style="background-color: #ccfff2">Stemming</span>: Apply a stemmer on the paragraphs, so that inflected forms are mapped to the base form. For example, for Python the popular natural language toolkit [nltk](http://www.nltk.org/howto/stem.html) has an easy to use stemmer. In case you are using R, you can try the [Snowball stemmer](https://www.rdocumentation.org/packages/corpus/versions/0.10.2/topics/stem_snowball). You can find out how to install nltk from [here](https://www.nltk.org/install.html). It will take a while to run! So, grab a coffee and wait :D

Finally, after stemming: 

*'wrote review manufactur contact explain current pictur amazon phone vertic stand phone horizont'*

In [20]:
from nltk.stem import PorterStemmer
stemmer = PorterStemmer()
texts_stem = texts_cleaned.apply(lambda _: ' '.join([stemmer.stem(word) for word in _.split()]))
print(texts_stem)

0          wrote review manufactur contact explain use in...
1          suck bare pick definit car pretti don t wast m...
2               write short blew 2 fuse car appar demand 12v
3          absolut memori buy m go review goe away review...
4                                                 ok doe job
                                 ...                        
1711514                                 fast ship work great
1711515    fit isn t great relay hard press connector mechan
1711516    bought auxiliari horn car need run relay guy f...
1711517    way better stock nice bright white super easi ...
1711518    great product fast ship awesom commun product ...
Name: reviewText, Length: 1711519, dtype: object


5. Finally, filter the data by selecting reviews where the field <span style="background-color: #ccfff2">overall</span> is 4 or 5, and store the review texts in a file named <span style="background-color: #ccfff2">pos.txt</span>. Similarly, select reviews with rating 1 or 2 and store them in a file named <span style="background-color: #ccfff2">neg.txt</span>. Ignore the reviews with overall rating 3. Each line in the two files should contain exactly one preprocessed review text without the rating.

In [21]:
pos_review = texts_stem[df['overall'].isin([4, 5])]
neg_review = texts_stem[df['overall'].isin([1, 2])]
pos_review.to_csv('pos.txt', index=False)
neg_review.to_csv('neg.txt', index=False)

## Exercise 3 | SQL basics

Next, let's take a refresher on the basics of SQL. In this exercise, you will be working on the simplified Northwind 2000 SQLite database. You can download the database from Kaggle [here](https://courses.mooc.fi/api/v0/files/course/f92ffc32-2dd4-421d-87f3-c48800422cc5/files/VEKX2bxGCDGyojG902gmYZTXCnrAQw.zip).

To test your SQL queries and complete the exercise, you can download and install SQLite if you don't yet have it installed.

Please write SQL queries for the tasks on the simplified Northwind 2000 SQLite database.

1. List the first name, last name, and hire date of all employees hired after January 1st, 1994.

    SELECT FirstName,LastName, HireDate FROM Employees WHERE HireDate > '1994-01-01'; 


2. Count how many orders each customer has placed. 

    SELECT CustomerID, COUNT(*) FROM Orders GROUP BY CustomerID;


3. Find the names of all customers who have ordered the product "Chai".

    SELECT DISTINCT c.CompanyName FROM Customers c 
    JOIN Orders o ON c.CustomerID = o.CustomerID 
    JOIN OrderDetails od ON o.OrderID = od.OrderID 
    JOIN Products p ON od.ProductID = p.ProductID WHERE p.ProductName = 'Chai';


4. Find all orders that have been placed but not yet shipped. 

    SELECT * FROM Orders WHERE ShippedDate IS NULL



5. Find the customer who has placed the most orders.

    SELECT CustomerID,COUNT(*) FROM Orders GROUP BY CustomerID ORDER BY COUNT(*) DESC LIMIT 1
    

In [23]:
import sqlite3
conn = sqlite3.connect('northwind2000-simplified.sqlite')
print(conn.execute("SELECT name FROM sqlite_master WHERE type='table'").fetchall())
print(conn.execute("SELECT FirstName,LastName, HireDate FROM Employees WHERE HireDate > '1994-01-01'").fetchall())
print(conn.execute("SELECT CustomerID,COUNT(*) FROM Orders GROUP BY CustomerID").fetchall())
print(conn.execute("SELECT DISTINCT c.CompanyName FROM Customers c JOIN Orders o ON c.CustomerID = o.CustomerID JOIN OrderDetails od ON o.OrderID = od.OrderID JOIN Products p ON od.ProductID = p.ProductID WHERE p.ProductName = 'Chai'").fetchall())
print(conn.execute("SELECT * FROM Orders WHERE ShippedDate IS NULL").fetchall())
print(conn.execute("SELECT CustomerID,COUNT(*) FROM Orders GROUP BY CustomerID ORDER BY COUNT(*) DESC LIMIT 1").fetchall())
conn.close()


[('Categories',), ('sqlite_sequence',), ('Customers',), ('Employees',), ('OrderDetails',), ('Orders',), ('Products',), ('Shippers',), ('Suppliers',)]
[('Robert', 'King', '1994-01-02'), ('Laura', 'Callahan', '1994-03-05'), ('Anne', 'Dodsworth', '1994-11-15')]
[('ALFKI', 6), ('ANATR', 4), ('ANTON', 7), ('AROUT', 13), ('BERGS', 18), ('BLAUS', 7), ('BLONP', 11), ('BOLID', 3), ('BONAP', 17), ('BOTTM', 14), ('BSBEV', 10), ('CACTU', 6), ('CENTC', 1), ('CHOPS', 8), ('COMMI', 5), ('CONSH', 3), ('DRACD', 6), ('DUMON', 4), ('EASTC', 8), ('ERNSH', 30), ('FAMIA', 7), ('FOLIG', 5), ('FOLKO', 19), ('FRANK', 15), ('FRANR', 3), ('FRANS', 6), ('FURIB', 8), ('GALED', 5), ('GODOS', 10), ('GOURL', 9), ('GREAL', 11), ('GROSR', 2), ('HANAR', 14), ('HILAA', 18), ('HUNGC', 5), ('HUNGO', 19), ('ISLAT', 10), ('KOENE', 14), ('LACOR', 4), ('LAMAI', 14), ('LAUGB', 3), ('LAZYK', 2), ('LEHMS', 15), ('LETSS', 4), ('LILAS', 14), ('LINOD', 12), ('LONEP', 8), ('MAGAA', 10), ('MAISD', 7), ('MEREP', 13), ('MORGK', 5), ('NO

**Remember to submit your solutions. You can return this Jupyter notebook (.ipynb) or .py, .R, etc depending on your programming preferences. Remember to also submit your SQL queries. No need to submit the text files for the programming exercises.**